parse.py

```diff
@@ -0,0 +1,157 @@
+import os
+import re
+import sys
+import time
+import docopt
+import numpy as np
+import pickle
+
+import config
+import database
+from utils import *
+
+import mwapi
+from mwapi.errors import APIError
+import mwparserfromhell
+
+from keras.models import load_model
+from keras.preprocessing.sequence import pad_sequences
+from keras import backend as K
+K.clear_session()
+K.set_session(K.tf.Session(config=K.tf.ConfigProto(intra_op_parallelism_threads=10, inter_op_parallelism_threads=10)))
+
+cfg = config.get_localized_config()
+WIKIPEDIA_BASE_URL = 'https://' + cfg.wikipedia_domain
+
+def load_citation_needed():
+    # load the vocabulary and the section dictionary
+    vocab_w2v = pickle.load(open(cfg.vocb_path, 'rb'))
+    section_dict = pickle.load(open(cfg.section_path, 'rb'), encoding='latin1')
+
+    # load Citation Needed model
+    model = load_model(cfg.model_path)
+    return model, vocab_w2v, section_dict
+
+def run_citation_needed(sentences, model, vocab_w2v, section_dict):
+    max_len = model.input[0].shape[1].value
+    # construct the training data
+    X = []
+    sections = []
+
+    for text, _, section in sentences:
+        # handle abbreviation, special characters
+        # transform the text into a word list
+        wordlist = text_to_word_list(text)
+
+         # construct the word vector for word list
+        X_inst = []
+        for word in wordlist:
+            if max_len != -1 and len(X_inst) >= max_len:
+                break
+            X_inst.append(vocab_w2v.get(word, vocab_w2v['UNK']))
+
+        X.append(X_inst)
+        sections.append(section_dict.get(section, 0))
+
+    # pad all word vectors to max_len
+    X = pad_sequences(X, maxlen=max_len, value=vocab_w2v['UNK'], padding='pre')
+    sections = np.array(sections)
+
+    return model.predict([X, sections])
+
+def extract(wikitext):
+    sentences = []
+    paragraphs = {}
+
+    for section in wikitext.get_sections(levels=[2], include_lead=True):
+        if not section.filter_headings(): # no heading -> is lead section
+            section_name = 'MAIN_SECTION'
+            headings = []
+
+        elif section.filter_headings()[0].title.strip() in cfg.sections_to_skip:
+            # ignore sections which content dont need citations
+            continue
+
+        else:
+            continue # for test, extract function needs to be improved
+            section_name = section.filter_headings()[0].title.strip()
+            # store all (sub)heading names
+            headings = [h.title for h in section.filter_headings()]
+
+        # split section content into paragraphs
+        section_parag = re.split("\n+", section.strip_code())
+
+        for i, paragraph in enumerate(section_parag):
+            pid = mkid(section_name+str(i))
+            paragraphs[pid] = paragraph
+            # clean hyperlinks which strip_code() did not remove
```

Comment:
Do you have an example where this happens? I think it should not happen in most cases according to the documentation for [strip_code](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.strip_code), but we can also process links individually if we have to with [filter_external_links](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.filter_external_links).

Also, we should check with @mirrys whether we really can't just keep the links. As I understand it, the model only has a fixed set of words that it uses for predictions, so anything not in that list (including links) will just be ignored later on anyway...

Reply:
For example, the first paragraph in "[Business](https://en.wikipedia.org/wiki/Business)", strip_code doesn’t remove the content between \<ref\> and \</ref\> which possibly contains text and hyperlinks. But I just found I can use [filter_tags](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.filter_tags) to catch them. 

The text returned by strip_code are the following which bold text should be removed: 

> Business is the activity of making one's living or making money by producing or buying and selling products (such as goods and services). **Compare: American Heritage Dictionary "business [:] 1. The activity of buying and selling  commodities, products, or services".** **Longman Business English Dictionary** **Longman Dictionary of Contemporary English "business [:] 1 [...] the activity of making money by producing or buying and selling goods, or providing services".** **Oxford Living Dictionaries "business [:] 2 The practice of making one's living by engaging in commerce."** Simply put, it is "any activity or enterprise entered into for profit. It does not mean it is a company, a corporation, partnership, or have any such formal organization, but it can range from a street peddler to General Motors."**Burton's Legal Thesaurus, 4E. S.v. "business." Retrieved April 1, 2018 from https://legal-dictionary.thefreedictionary.com/business** 

And I don't think we can keep these links and texts. Because for those words which are not in the list, the model fills `'UNK'` instead of ignoring them.

Reply:
Ah yes, I guess it makes sense that `strip_code` doesn't remove tags.

But how does `'UNK'` affect the output of the model? Does their presence actually change the score for a sentence?

Reply:
I think because the input for the RNN model is a fixed length of the word vectors, if we don't remove extra hyperlinks, they become many `UNK` filling the word vectors. When the length exceeds the limitation of maximum word vectors, they are truncated and some meaningful words are missing.

Reply:
Wow, that's interesting. It seems a little bit odd to me that an unknown word can push known words out of the vector if it appears in the beginning of the sentence, but maybe that's by design -- I'll have to defer to @mirrys here.

Reply:
> I think because the input for the RNN model is a fixed length of the word vectors, if we don't remove extra hyperlinks, they become many `UNK` filling the word vectors. When the length exceeds the limitation of maximum word vectors, they are truncated and some meaningful words are missing.

@AikoChou is right, this is a limitation of the model design. But my understanding the maximum length for the fixed sentence has been calculated to be the largest sentence in the training set -- let me check with my co-author and get back to this.

More in general, because token are treated as words when present in the dictionary, or ar unknown when not recognized,  I would probably just include links as well and leave the word analysis process as close as possible to the original script.

Reply:
From my perspective, the links are unnecessary to keep because they belong to citations in this case, but we should keep the links if they show in the content.